File: concepts/06_text_binary_and_serialization/03_regular_expressions_and_state/lua/test_02_zero_width_and_failure_boundaries.lua
-- Common question: how do zero-width matches and invalid patterns terminate?
-- Inputs: frontier patterns, empty captures, malformed classes, and gsub callbacks.
-- Observations: boundary positions, finite advancement, syntax errors, and callback failure propagation.
-- polyglot-family: text_binary_and_serialization
-- polyglot-concept: regular_expressions_and_state
-- polyglot-related: languages/lua/standard_library/09_strings_patterns_utf8/
-- polyglot-related+: test_067_pattern_language_iteration_and_substitution.lua

local t = require("support.assertions")

local start_position, end_position = string.find(" a", "%f[%a]")
t.equal(start_position, 2)
t.equal(end_position, 1)

local positions = {}
for position in string.gmatch("ab", "()") do
    positions[#positions + 1] = position
end
t.equal(table.concat(positions, ","), "1,2,3")

t.raises(function() string.match("abc", "[") end)
t.raises(function()
    string.gsub("a", ".", function() error("replacement failed") end)
end, "replacement failed")

t.done()

File: concepts/06_text_binary_and_serialization/03_regular_expressions_and_state/ruby/test_02_zero_width_and_failure_boundaries.rb
# 共同问题：零宽匹配怎样前进，无效 pattern 和替换 callback 失败怎样传播。
# 输入：anchors、lookahead、空 pattern、无效源码和 gsub block；观察：有限迭代、RegexpError 与原异常。
# polyglot-family: text_binary_and_serialization
# polyglot-concept: regular_expressions_and_state
# polyglot-related: languages/ruby/language/07_strings_regex_and_patterns/
# polyglot-related+: test_052_regexp_captures_substitution_and_match_state.rb

require "assertions"

A = PolyglotAssertions

A.equal(["", "", ""], "ab".scan(//))
A.equal(["", ""], "ab".scan(/(?=[ab])/))
A.equal("XaXbX", "ab".gsub(//, "X"))
A.raises(RegexpError) { Regexp.new("[") }

error = RuntimeError.new("replacement")
captured = A.raises(RuntimeError, "replacement") do
  "a1".gsub(/\d/) { raise error }
end
A.same(error, captured)

A.done
